src/Api/Upload.php

```diff
@@ -0,0 +1,207 @@
+			$max[] = Str::toBytes('100M');
+		}
+
+		// to be sure, only use 95% of the max possible upload size
+		return (int)floor(min($max) * 0.95);
+	}
+
+	/**
+	 * Clean up tmp directory of stale files
+	 */
+	public static function clean(): void
+	{
+		foreach (Dir::files($dir = static::tmp(), [], true) as $file) {
+			// remove any file that hasn't been altered
+			// in the last 24 hours
+			if (F::modified($file) < time() - 86400) {
+				F::remove($file);
+			}
+		}
+
+		// remove tmp directory if completely empty
+		if (Dir::isEmpty($dir) === true) {
+			Dir::remove($dir);
+		}
+	}
+
+	/**
+	 * Returns root of directory used for
+	 * temporarily storing (incomplete) uploads
+	 * @codeCoverageIgnore
+	 */
+	protected static function tmp(): string
+	{
+		return App::instance()->root('cache') . '/.uploads';
+	}
+
+	/**
+	 * Ensures the sent chunk is valid
+	 *
+	 * @throws \Kirby\Exception\DuplicateException Duplicate first chunk (same filename and id)
+	 * @throws \Kirby\Exception\Exception Chunk offset does not match existing tmp file
+	 * @throws \Kirby\Exception\InvalidArgumentException The maximum file size for this blueprint was exceeded
+	 * @throws \Kirby\Exception\NotFoundException Subsequent chunk has no  existing tmp file
+	 */
+	protected static function validateChunk(
+		string $source,
+		string $tmp,
+		int $total,
+		int $offset,
+		string|null $template = null
+	): void {
+		$name   = basename($tmp);
+		$parent = new Page(['slug' => 'tmp']);
+		$file   = new File([
+			'filename' => $name,
+			'parent'   => $parent,
+			'template' => $template
+		]);
+
+		// if the blueprint `maxsize` option is set,
+		// ensure that the total size communicated in the header
+		// as well as the current tmp size after adding this chunk
+		// does not exceed the max limit
+		if (
+			($max = $file->blueprint()->accept()['maxsize'] ?? null) &&
+			(
+				$total > $max ||
+				(F::size($source) + F::size($tmp)) > $max
+			)
+		) {
+			throw new InvalidArgumentException(['key' => 'file.maxsize']);
+		}
+
+		if (F::exists($tmp) === true) {
+			// sent chunk is expected to be the first part,
+			// but tmp file already exists
+			if ($offset === 0) {
+				throw new DuplicateException('A tmp file upload with the same filename and upload id already exists: ' . $name);
+			}
+
+			// sent chunk's offset is not the continuation of the tmp file
+			if ($offset !== F::size($tmp)) {
+				throw new Exception('Chunk offset ' . $offset . ' does not match the existing tmp upload file size of ' . F::size($tmp));
+			}
+		} elseif ($offset > 0) {
+			// sent chunk is not expected to be the first part,
+			// but no tmp file to continue exists yet
+			throw new NotFoundException('Chunk offset ' . $offset . ' for non-existing tmp file: ' . $name);
+		} elseif ($offset === 0) {
+			// validate file (extension, name) for first chunk;
+			// will also be validate again by `$model->createFile()`
+			// when completely uploaded
+			FileRules::validFile($file, false);
+		}
```

Comment:
To get rid of the `elseif`s:

```suggestion

			// existing file is valid, done validating
			return;
		}

		if ($offset === 0) {
			// validate file (extension, name) for first chunk;
			// will also be validate again by `$model->createFile()`
			// when completely uploaded
			FileRules::validFile($file, false);
		}

		throw new NotFoundException('Chunk offset ' . $offset . ' for non-existing tmp file: ' . $name);
```

Reply:
Ended up fixing it differently as this solution would always throw the last exception for valid chunks. But managed to untangle the if-elseifs a bit.